### PR TITLE
perf: Rand index computation

### PR DESCRIPTION
**Pull request by Lucas-Prates**
Greetings. 

I believe there is a minor mistake on the scaling done on the **hamming** method, it should be _n_samples*(n_samples-1)_.

I also provided an additional implementation of the Rand index under the method **randindex_cpd**. This implementation is based on an specific expression for the metric on change-point problems. If we have _N_ samples and change-point sets of size _r_ and _s_, this algorithm runs on _O(r+s)_ in time and _O(1)_ in memory complexity. The traditional implementation runs on _O(rs+N)_ in time and _O(rs)_ in memory, albeit your implementation might use less due to sparsity. This new implementation seems to perform some orders of magnitude better, depending on _N_, in a few profiling that I did.

I can provide the proof of correctness of the algorithm if requested.

Thanks.

### CODE REVIEW

**Comment by deepcharles** [comment]
Many thanks for this very useful PR.

The implementation of Rand index is indeed quite naive. I will take a look.

**Comment by deepcharles** [comment]
Am I correect in assuming that the code you provided implements [this report](https://arxiv.org/pdf/2112.03738.pdf)? I am still looking at the pseudo-code.

File: src/ruptures/metrics/randindex.py [hunk 1]
```diff
@@ -14,3 +15,41 @@ def randindex(bkps1, bkps2):
...
+        bkps1 (list): sorted list of the last index of each regime.
+        bkps2 (list): sorted list of the last index of each regime.
+
+    Returns:
+        float: Rand index
+    """
+    sanity_check(bkps1, bkps2)
+    n_samples = max(bkps1)
+    disagreement = 0
+    bkps1.insert(0, 0)
```
Comment (deepcharles) [inline src/ruptures/metrics/randindex.py:34]: Do not modify in-place variables that can be used outside the function
```suggestion
```

File: src/ruptures/metrics/randindex.py [hunk 2]
```diff
@@ -14,3 +15,41 @@ def randindex(bkps1, bkps2):
...
+    bkps2.insert(0, 0)
```
Comment (deepcharles) [inline src/ruptures/metrics/randindex.py:35]: Do not modify in-place variables that can be used outside the function
```suggestion
```

File: src/ruptures/metrics/randindex.py [hunk 3]
```diff
@@ -14,3 +15,41 @@ def randindex(bkps1, bkps2):
...
+            disagreement += nij * abs(bkps1[i + 1] - bkps2[j + 1])
+
+            # we can skip the rest of the iteration, nij will be 0
+            if bkps1[i + 1] < bkps2[j + 1]:
+                break
+
+            else:
+                beginj = j + 1
+
+    bkps1.pop(0)
```
Comment (deepcharles) [inline src/ruptures/metrics/randindex.py:52]: Do not modify in-place variables that can be used outside the function
```suggestion
```

File: src/ruptures/metrics/randindex.py [hunk 4]
```diff
@@ -14,3 +15,41 @@ def randindex(bkps1, bkps2):
...
+    bkps2.pop(0)
```
Comment (deepcharles) [inline src/ruptures/metrics/randindex.py:53]: Do not modify in-place variables that can be used outside the function
```suggestion
```

File: src/ruptures/metrics/randindex.py [hunk 5]
```diff
@@ -14,3 +15,41 @@ def randindex(bkps1, bkps2):
...
+        float: Rand index
+    """
+    sanity_check(bkps1, bkps2)
+    n_samples = max(bkps1)
+    disagreement = 0
+    bkps1.insert(0, 0)
+    bkps2.insert(0, 0)
+
+    beginj = 0  # avoids unnecessary computations
+    for i in range(len(bkps1) - 1):
```
Comment (deepcharles) [inline src/ruptures/metrics/randindex.py:38]: Try to use more informative variable names (not `i` or `j`)
```suggestion
    for (start1, end1) in pairwise([0]+bkps1):
```

(`pairwise` is available in `ruptures.utils`.)

File: src/ruptures/metrics/randindex.py [hunk 6]
```diff
@@ -14,3 +15,41 @@ def randindex(bkps1, bkps2):
...
+        for j in range(beginj, len(bkps2) - 1):
```
Comment (deepcharles) [inline src/ruptures/metrics/randindex.py:39]: Try to use more informative variable names (not `i` or `j`)
```suggestion
    for (start2, end2) in pairwise([0]+bkps2):
```

(`pairwise` is available in `ruptures.utils`.)

File: src/ruptures/metrics/randindex.py [hunk 7]
```diff
@@ -14,3 +15,41 @@ def randindex(bkps1, bkps2):
...
+            nij = min(bkps1[i + 1], bkps2[j + 1])
```
Comment (deepcharles) [inline src/ruptures/metrics/randindex.py:40]:
```suggestion
            nij = min(end1, end2)
```

File: src/ruptures/metrics/randindex.py [hunk 8]
```diff
@@ -14,3 +15,41 @@ def randindex(bkps1, bkps2):
...
+            nij -= max(bkps1[i], bkps2[j])
```
Comment (deepcharles) [inline src/ruptures/metrics/randindex.py:41]:
```suggestion
            nij -= max(start1, start2)
```

File: src/ruptures/metrics/randindex.py [hunk 9]
```diff
@@ -14,3 +15,41 @@ def randindex(bkps1, bkps2):
...
+            nij = max(nij, 0)
+            disagreement += nij * abs(bkps1[i + 1] - bkps2[j + 1])
```
Comment (deepcharles) [inline src/ruptures/metrics/randindex.py:43]:
```suggestion
            disagreement += nij * abs(end1 - end2)
```

File: src/ruptures/metrics/randindex.py [hunk 10]
```diff
@@ -14,3 +15,41 @@ def randindex(bkps1, bkps2):
...
+
+            # we can skip the rest of the iteration, nij will be 0
+            if bkps1[i + 1] < bkps2[j + 1]:
```
Comment (deepcharles) [inline src/ruptures/metrics/randindex.py:46]:
```suggestion
            if end1 < end2:
```

File: src/ruptures/metrics/randindex.py [hunk 11]
```diff
@@ -14,3 +15,41 @@ def randindex(bkps1, bkps2):
         float: Rand index
     """
     return 1 - hamming(bkps1, bkps2)
+
+
+def randindex_cpd(bkps1, bkps2):
```
Comment (deepcharles) [inline src/ruptures/metrics/randindex.py:20]: Do not create another function, replace `randindex`.

File: src/ruptures/metrics/randindex.py (hunk 6 above)
Comment (deepcharles) [reply src/ruptures/metrics/randindex.py:39]: In order to skip unnecessary computations (using `beginj`), replace `pairwise([0]+bkps2)` by `islice(pairwise([0]+bkps2), beginj, None)` (`islice` is available from the `itertools` library).

**Comment by Lucas-Prates** [comment]
> Am I correct in assuming that the code you provided implements [this report](https://arxiv.org/pdf/2112.03738.pdf)? I am still looking at the pseudo-code.

Yes, I am the author of the report and the code is an implementation of that pseudo-code. Thank you for all your observations/corrections, I apologize for the mess. The implementation using `islice` is much cleaner, I tried using pairwise but did not know how to skip iterations.

**Comment by Lucas-Prates** [comment]
For the sake of completeness, I will provide an observation on the run time of the algorithms.

I did a simple simulation, and the figure below shows a comparison of their performance as sample size increases. _ri_clean_ denotes the algorithm after the corrections, _ri_hotmess_ the algorithm from the initial PR and _ri_ham_  the original implementation.


runtime_comp


For large sample sizes, e.g. _N > 1e5_, _ri_clean_  takes much more time than _ri_hotmess_, which I do not understand why. I did some profiling, and both algorithms are almost identical: they perform the exact same number of iterations on the loops, and their memory usage (as shown by the profiler) is the same. My guess to explain this discrepancy is on the evaluation of `[0] + bkps1` and `[0] + bkps2` on the argument of the `pairwise` function. Is python copying the list? If so, other methods using this construction might suffer from the same problem? But of course, this last part is speculation.

**Comment by deepcharles** [comment]
Thanks for this detailed information. Indeed, this is surprising to see that `ri_clean` performs so much worse.

Can you tell us if using temporary variables (e.g. `bkps1_with_0 = [0] + bkps1`) solves the problem? Also, is `n_samples` the size of `bkps1`/`bkps2`?

File: src/ruptures/metrics/__init__.py [hunk 12]
```diff
@@ -20,4 +20,4 @@
 from .timeerror import meantime
 from .precisionrecall import precision_recall
 from .hamming import hamming
-from .randindex import randindex
+from .randindex import randindex, randindex_cpd
```
Comment (deepcharles) [inline src/ruptures/metrics/__init__.py:23]:
```suggestion
from .randindex import randindex
```

File: tests/test_metrics.py [hunk 13]
```diff
@@ -1,6 +1,12 @@
 import pytest
 
-from ruptures.metrics import hausdorff, meantime, precision_recall, randindex
+from ruptures.metrics import (
+    hausdorff,
+    meantime,
+    precision_recall,
+    randindex,
+    randindex_cpd,
```
Comment (deepcharles) [inline tests/test_metrics.py:8]:
```suggestion

```

**Comment by Lucas-Prates** [comment]
I did try to use temporary variables, but the effect seems to persist, as you can see below

runtime_comparison_w0

_ri_clean_w0_ refers to using temporary variables. I also omitted the original implementation.

With respect to your question about `n_samples`, if you are referring to the variable on the code, no, it is the number of samples of the data set, which should be the maximum (and last entry) of `bkps1` and `bkps2`.

However, I believe your doubt is on the sample size of the simulation. If the sample size was _N_ (which I varied from 1e3 to 1e4 in steps of 1e3), the change-point sets sizes were _N/2_. I did this because the original implementation allocated a _N*N_ matrix independently of the change-point set, so I wanted to see the effect. On the figure of this comment, I displayed the information directly on the change-point set.

File: src/ruptures/metrics/randindex.py [hunk 14]
```diff
@@ -1,16 +1,38 @@
...
-        bkps2 (list): list of the last index of each regime.
+        bkps1 (list): sorted list of the last index of each regime.
+        bkps2 (list): sorted list of the last index of each regime.
 
     Returns:
         float: Rand index
     """
-    return 1 - hamming(bkps1, bkps2)
+    sanity_check(bkps1, bkps2)
+    n_samples = max(bkps1)
```
Comment (deepcharles) [inline src/ruptures/metrics/randindex.py:19]:
```suggestion
    n_samples = bkps1[-1]
```

**Comment by deepcharles** [comment]
I have made some changes. Can you check if the performance is still so-so?

**Comment by Lucas-Prates** [comment]
Yes, I will check and post it here when I get the chance.

**Comment by Lucas-Prates** [comment]
Sorry for the delay.

It seems to have improved with the last changes, but there is still a gap.

runtime_comparison

I ran the simulations on my machine. I can provide the scripts if you want to test it yourself.

**Comment by oboulant** [comment]
Thx @Lucas-Prates ! Yes, can you please provide the script ?

**Comment by Lucas-Prates** [comment]
Sure! Just make a venv, install the requirements and run "run_all.sh" ( or the files separately). It should produce a pickle and csv from the simulation results, along with a .png figure for the runtime comparison.
[simulation_ruptures.zip](https://github.com/deepcharles/ruptures/files/7802507/simulation_ruptures.zip)

The "time profiling" here is very simplified, but I think that does not account for the discrepancy.

**Comment by oboulant** [comment]
I looked from several angles and at the end, it seems to me that it is really the impact of using iterables for which we have to unpack values. 
In other words, maybe there is not much that would more efficient than accessing an item in a list with its index. 

Adding the `pairwise` function adds an overhead (which is not the reason for the spread between `hotmess` and `clean`, but still), but more importantly it seems like consequently, the iteration over the iterable returned by `pairwise` slows the process as compared to the iteration over `range` because in the former case the system is busy doing stuffs unrelated to the computation performed within the loop. 

See the figures below : 

1. The `hotmess` implementation stays busy doing computations related to what is performed in the inner loop (`min`, `max`, `abs`)
Screenshot 2022-01-05 at 23 13 11

2. The `clean` implementation is busy doing things unrelated to what is performed in the inner loop
Screenshot 2022-01-05 at 23 12 31

The time scales on the x-axis on both graph are the same !

I tried to save some calls to `pairwise` by calling it once for `bkps1` and once for `bkps2`, same behaviour ⬇️ 

1. `hotmess` 
Screenshot 2022-01-05 at 23 36 58

2. `clean`

Screenshot 2022-01-05 at 23 36 14

I hope this is clear, I can reformulate if necessary. 

My opinion, as the light of this, and if I did not miss anything, would be to stick to the `hotmess` implementation that scales. And TBH, IMO it is not not clean, it is clean :)

**Comment by deepcharles** [comment]
Thanks @oboulant for the detailed comment.
Since the call to `pairwise` in the second for loop is costly, the best approach is to create a list to iterate over, i.e. replace

**Comment by oboulant** [comment]
> Thanks @oboulant for the detailed comment. Since the call to `pairwise` in the second for loop is costly, the best approach is to create a list to iterate over, i.e. replace `pairwise(bkps2_with_0)` by `list_of_regimes2` as in [#222 (comment)](https://github.com/deepcharles/ruptures/pull/222#discussion_r779195252) where `list_of_regimes2=list(pairwise(bkps2_with_0))`.
> 
> This solves the complexity discrepancy between the two versions, and keeps the code clean.

I agree with the proposed replacement in the sense that it is cleaner because it saves a call to `pairwise` each time we start over with the inner loop. 
Nevertheless, it is basically what I did in the second set of screenshots that I shared : it does not solves the performance issue. 

At the end, seems like iterating over `range` vs. over what we have for `list_of_regimes2` is way more efficient. That is why I said why not go with the implementation with `range` ?

**Comment by Lucas-Prates** [comment]
That is a very interesting/precise analysis @oboulant, and it seems to explain the discrepancy we have been discussing. Just to check my understanding, the blank gaps in the figures associated with `clean` are due to computations performed outside the inner loop?

**Comment by oboulant** [comment]
Blank gaps corresponds to time spent **directly** in `randindex()` and not in a child call. 

Since the code running is the following  : 

```python
    for (start1, end1) in pairwise(bkps1_with_0):
        for (start2, end2) in pairwise(bkps2_with_0[beginj:]):
            nij = max(min(end1, end2) - max(start1, start2), 0)
            disagreement += nij * abs(end1 - end2)

            # we can skip the rest of the iteration, nij will be 0
            if end1 < end2:
                break
            else:
                beginj += 1
```

Blank gaps means we are performing task that are **not** : 
* `max` 
* `min` 
* `abs`

So basically it leaves that blank spaces are computations related to:
* `+=` operations
* `*` operations
* `<` operations
* iteration and unpacking of the iterables

Since the `+=`, `*`, `<` operations are also performed in the `hotmess`, it leaves the thing on the iterables. But maybe I am missing something !

**Comment by Lucas-Prates** [comment]
Thank you for the detailed explanation. I do not know what could be missing.

**Comment by deepcharles** [comment]
Now, there should not be any difference in performance between the first version and the current one.

I also cleaned the associated documentation and added a reference to your work @Lucas-Prates .

We should be good to go now.

**Comment by oboulant** [comment]
runtime_comparison

Here you are, with the discussed modifications

**Pull request merged**